Add TA-131639 (#18981)
cockroachdb/docs: src/current/advisories/a131639.md

---
title: Technical Advisory 131639
advisory: A-131639

> summary: During a sustained period of disk slowness in the presence of lease transfers, it is possible for some writes in a transaction that straddles multiple ranges to be lost.

Author: ✅

toc: true
affected_versions: v23.1.0 to v23.1.26, v23.2.0 to v23.2.10, v24.1.0
advisory_date: 2024-10-03
docs_area: releases
---

Publication date: {{ page.advisory_date | date: "%B %e, %Y" }}

## Description

> Since v22.2, through v24.2, in circumstances where there is a sustained period of disk slowness in the presence of lease transfers, it is possible for some writes in a transaction that straddles multiple ranges to be lost. In cases where these lost writes are not also present on one or more secondary indexes that could allow for reconstructing or repairing the row, these writes could be irrecoverably lost.

Author: ✅

However, in certain scenarios, these writes may be fully recoverable. For example, if the lost writes were on a secondary index, these could be fully repaired from the primary index, or a primary index could be fully repaired if there exists a set of secondary indexes that fully covers the primary index columns. The lost writes may also be partially recoverable. For example, if only some of the lost writes from a primary index are contained in secondary indexes.

> In detail, since v22.2, when a lease for a range is transferred, it is done so as an expiration-based lease. This lease is then promoted to an epoch-based lease. In situations where the block device backing a store is experiencing intermittent slowness over an extended period of time, the expiration time for the lease transferred to the node with the slow stores can move back in time during this promotion. The lease expiration is said to have "regressed". This lease expiration regression can result in a brief period of time whereby another node can claim the lease, and the range can have two leaseholders.

Author: ✅

> While transactions will typically be routed from gateway nodes to the true leaseholder (on the node with the healthy store), and write intents will be written on this true leaseholder, on the node with the slow store these write intents are still queued up in the raft log, and have not been applied. If async intent resolution (the means by which intents are committed to durable records, once a transaction commits) is subsequently triggered on this slow node before it has caught up on its log, intent resolution becomes a no-op, instead of marking the intent as committed. A future transaction that interacts with the row may then come across this intent, and if the committed transaction's record has been GC-ed at that point, it may incorrectly determine that the intent belongs to an aborted transaction. This would then cause the intent to be cleaned up, effectively dropping the committed write.

Author: ✅

> While transactions will typically be routed from gateway nodes to the true leaseholder (on the node with the healthy store), and write intents will be written on this true leaseholder, on the node with the slow store these write intents are still queued up in the raft log, and have not been applied. If async intent resolution (the means by which intents are committed to durable records, once a transaction commits) is subsequently triggered on this slow node before it has caught up on its log, intent resolution becomes a no-op, instead of marking the intent as committed. A future transaction that interacts with the row may then come across this intent, and if the committed transaction's record has been GC-ed at that point, it may incorrectly determine that the intent belongs to an aborted transaction. This would then cause the intent to be cleaned up, effectively dropping the committed write.

Author: thx ✅ 

 also checked one other use of this phrase and updated it to capitalize its use of 'Raft'

> While transactions will typically be routed from gateway nodes to the true leaseholder (on the node with the healthy store), and write intents will be written on this true leaseholder, on the node with the slow store these write intents are still queued up in the range's Raft log, and have not been applied. If async intent resolution (the means by which intents are committed to durable records, once a transaction commits) is subsequently triggered on this slow node before it has caught up on its Raft log, intent resolution becomes a no-op, instead of marking the intent as committed. A future transaction that interacts with the row may then come across this intent, and if the committed transaction's record has been GC-ed at that point, it may incorrectly determine that the intent belongs to an aborted transaction. This would then cause the intent to be cleaned up, effectively dropping the committed write.

Reviewer: s/GC-ed/garbage collected/

Author: ✅

This bug is rare in that it requires intermittent slowness of a block device backing a store, coupled with lease transfers onto the store with the intermittently slow block device, followed by async intent resolution on the slow node before it has caught up on its raft log. Writes must touch more than one range. Single range writes, including those that benefit from the One-Phase-Commit (1PC) fast path are not affected.

All CockroachDB versions prior to and including the following are impacted by this bug:

- v23.1.0 - v23.1.26 and below
- v23.2.0 - v23.2.10 and below
- v24.1.0 - subsequent patch versions are unaffected

> Versions prior to 23.1 are no longer eligible for [maintenance support]({% link releases/release-support-policy.md %}), and this issue will not be fixed in those versions.

Author: ya let's do it, i was gonna say fixed is clearer but upon reflection addressed is probably the proper term in the sense that a patch addresses an issue but whether it fixes the issue is something empirical that can only be determined over time

## Statement

> This is resolved in CockroachDB by [#123442](https://github.com/cockroachdb/cockroach/pull/123442) which resolves an issue in the expiration-to-epoch lease promotion transition, where the lease's effective expiration could be allowed to regress, resulting in two nodes believing they are the leaseholder for a range.

Reviewer: ```suggestion
[#123442](https://github.com/cockroachdb/cockroach/pull/123442) resolves an issue in CockroachDB in the expiration-to-epoch lease promotion transition, where the lease's effective expiration could be allowed to regress, resulting in two nodes believing they are the leaseholder for a range.
```

Author: ✅

> The fix has been applied to maintenance releases of CockroachDB:

Author: nice totes agreed, "the patch" fits with the "addresses" language better

- [v23.1.27]({% link releases/v23.1.md%}#v23-1-27)
- [v23.2.11]({% link releases/v23.2.md%}#v23-2-11)
- [v24.1.1]({% link releases/v24.1.md%}#v24-1-1)

This public issue is tracked by [131639](https://github.com/cockroachdb/cockroach/issues/131639).

## Mitigation

> Users are encouraged to upgrade to [v23.1.27]({% link releases/v23.1.md%}#v23-1-27), [v23.2.11]({% link releases/v23.2.md%}#v23-2-11), [v24.1.1]({% link releases/v24.1.md%}#v24-1-1), or a later version that includes the fix.

Author: ✅

> ### Detection via logs

Reviewer 2: We now have a [script](https://github.com/cockroachlabs/support/pull/3101) that can be used, given a `debug.zip` as input.

The script requires two dependencies - `ripgrep` and `duckdb`. The install instructions for each can be found here:
- https://github.com/BurntSushi/ripgrep#installation
- https://duckdb.org/docs/installation

@rmloveland - thoughts on how to best include this? We could include the script inline, perhaps at the bottom of the TA, so as not to clutter things to much? Alternatively, we could host it somewhere, and reference it. Will defer to you.

Author: @nicktrav (and @nvanbenschoten as script author) I took a shot at including the script inline, and updated the section with an edited version of the description of https://github.com/cockroachlabs/support/pull/3101

PTAL

one thing i noticed is that the script outputs range IDs and the next section 'Detection via an audit of primary and secondary indexes' refers to checking indexes by name

is there guidance we can offer to users on how to get from the potentially affected range IDs to the names of indexes they need to check? 

AFAICT it's pretty easy to get from a schema object to a range ID with [`SHOW RANGES`](https://www.cockroachlabs.com/docs/v24.2/show-ranges.html) but i'm not sure if we have a supported (or unsupported) way to map from range IDs up to schema objects?  if we don't write it here, it seems folks will reach out to ask

Author: found some discussion [here](https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1716323894389819) of how to do approx a range ID -> SQL mapping and it looks pretty "internal" / unsupported / confusing but i defer to yall

Reviewer 3: This is a good question. Shaun M closes that thread with "The fact that a range can contain primary indexes, secondary indexes, and even have multiple tables within the same range - there’s a many-to-one relationship which is rough to untangle". That's pretty much right, and so it's not as clear an answer as it might seem to perform this mapping, at least not in an automated way. Sean C includes the kind of query we'd need to run to perform the mapping, but it's involved.

If we get to this point with a customer where the script detecting symptoms is pointing at range IDs, it might just be best to get support involved. The next steps are going to be different on a case-by-case basis.

Reviewer 4: Thanks, @nvanbenschoten. Given the various callouts we have to contact Support, can we leave the copy as-is, or is there anything you'd add to point them in the right direction about what to expect, even if the end result is needing to contact Support?

@nickelnickel can you also review this, please?

Reviewer 4: @nvanbenschoten: Nick (@nickelnickel) tells me he's good with this as-is, so if the copy's good to you, we are good to go.

Reviewer 3: I'm good with this as is.

While log lines alone are insufficient to confirm the occurrence of a lost write, they can be used to identify whether such a situation could have possibly occurred.

First, look for log lines that indicate intermittent slowness of a block device backing one or more stores. This can be seen with log lines like:

~~~
err=heartbeat failed on epoch increment
~~~

Followed by a log line like:

~~~
incremented n.* liveness epoch to .*
~~~

Next, during the same time period as the slowness, look for delayed lease promotions on the same node as the slow disk:

~~~
applied lease after .* replication lag, client traffic may have been delayed.*epo=.*prev.*exp=.*Request.*
~~~

> If all three of these log lines are present and the timestamp of the third log is less than the specified "replication lag" after the timestamp of the second log, then it is possible that a lost write occurred on the range identified in the third log line.

Author: > Maybe we can "bold" the bits of interesting (does markdown support bold within a code block??).

AFAICT the answer is no but we can do ASCII gymnastics like

```
INFO: the value of foo=bar at time t=123
                                ^^^^^^
                           LOOK FOR THIS
```

> ### Detection via an audit of primary and secondary indices

Reviewer: s/indices/indexes/
I know that "indices" is also correct, but it is not used the docs. "Indexes" is used throughout the docs instead.

Author: ✅

---
title: Technical Advisory 131639
advisory: A-131639
summary: During a sustained period of disk slowness in the presence of lease transfers, it is possible for some writes in a transaction that straddle multiple ranges to be lost.
toc: true

> affected_versions: v23.1.0 to v23.1.26, v23.2.0 to v23.2.10, v24.1.0

Reviewer 4: ```suggestion
affected_versions: v22.2, v23.1.0 to v23.1.26, v23.2.0 to v23.2.10, v24.1.0
```

adding 22.2 major version to match another recent TA

This bug is rare in that it requires intermittent slowness of a block device backing a store, coupled with lease transfers onto the store with the intermittently slow block device, followed by async intent resolution on the slow node before it has caught up on its Raft log. Writes must touch more than one range. Single range writes, including those that benefit from the One-Phase-Commit (1PC) fast path are not affected.

The following CockroachDB versions, as well as major versions prior to 23.1, are impacted by this bug:

> - v23.1.0 - v23.1.26

Reviewer 4: ```suggestion
- v22.2
- v23.1.0 - v23.1.26
```

Reviewer 2: LGTM

### Detection via logs

Users can run the script [a131639_check.sh](a131639_check.sh) to determine whether any lease expiration regressions are evident from their logs. The script must be run from within a decompressed [debug.zip]({% link v24.3/cockroach-debug-zip.md %}) directory.

As described previously, a lease expiration regression is only one of the two race conditions needed to encounter a lost write, with the other being a stale read during asynchronous intent resolution on the range that had the lease expiration regression. Consequently, this script does not provide a complete method of detection for a lost write. Running this script may result in a limited number of false positives. For complete confirmation, this script should be combined with the method described in [Detection via an audit of primary and secondary indexes](#detection-via-an-audit-of-primary-and-secondary-indexes).

Users should not run the script on a production system. Instead, users should move the `debug.zip` file elsewhere for analysis and run the script there.

> The script requires (and asks for) the following dependencies:

Reviewer 4: ```suggestion
The script requires the following dependencies:
```

This bug is rare in that it requires intermittent slowness of a block device backing a store, coupled with lease transfers onto the store with the intermittently slow block device, followed by async intent resolution on the slow node before it has caught up on its Raft log. Writes must touch more than one range. Single range writes, including those that benefit from the One-Phase-Commit (1PC) fast path are not affected.

> The following CockroachDB versions, as well as major versions prior to 23.1, are impacted by this bug:

Reviewer 4: ```suggestion
The following CockroachDB versions are impacted by this bug:
```

---
title: Technical Advisory 131639
advisory: A-131639
summary: During a sustained period of disk slowness in the presence of lease transfers, it is possible for some writes in a transaction that straddle multiple ranges to be lost.
toc: true
affected_versions: v23.1.0 to v23.1.26, v23.2.0 to v23.2.10, v24.1.0

> advisory_date: 2024-10-03

Reviewer 4: ```suggestion
advisory_date: 2024-10-07
```

docs_area: releases
---

Publication date: {{ page.advisory_date | date: "%B %e, %Y" }}

## Description

> Since v22.2, through v24.2, in circumstances where there is a sustained period of disk slowness in the presence of lease transfers, it is possible for some writes in a transaction that straddle multiple ranges to be lost. In cases where these lost writes are not also present on one or more secondary indexes that could allow for reconstructing or repairing the row, these writes could be irrecoverably lost.

Reviewer 5: ```suggestion
In v22.2 through v24.2, in circumstances where there is a sustained period of disk slowness in the presence of lease transfers, it is possible for some writes in a transaction that straddle multiple ranges to be lost. In cases where these lost writes are not also present on one or more secondary indexes that could allow for reconstructing or repairing the row, these writes could be irrecoverably lost.
```
Slight edit here because I think the "Since...through..." construct reads a little confusing. I'm confused about the distinction being made between "lost" and "irrevocably lost", and I suspect in the next paragraph, the first and second sentences are saying the same thing.

This bug is rare in that it requires intermittent slowness of a block device backing a store, coupled with lease transfers onto the store with the intermittently slow block device, followed by async intent resolution on the slow node before it has caught up on its Raft log. Writes must touch more than one range. Single range writes, including those that benefit from the One-Phase-Commit (1PC) fast path are not affected.

All CockroachDB versions prior to and including the following are impacted by this bug:

- v23.1.0 - v23.1.26 and below
- v23.2.0 - v23.2.10 and below

> - v24.1.0 - subsequent patch versions are unaffected

Reviewer 5: Would it be better to list impacted  v22.2 versions also, since that's the only other version that is impacted? So then it would be: v22.2.0-19, v23.1.0-26, v23.2.0-10, and v24.1.0.

Versions prior to 23.1 are no longer eligible for [maintenance support]({% link releases/release-support-policy.md %}), and this issue will not be addressed in those versions.

## Statement

> [#123442](https://github.com/cockroachdb/cockroach/pull/123442) resolves an issue in CockroachDB in the expiration-to-epoch lease promotion transition, where the lease's effective expiration could be allowed to regress, resulting in two nodes believing they are the leaseholder for a range.

Reviewer 3: @rmloveland do we prefer to point to Pull Requests from TAs, or to commits (in this case, https://github.com/cockroachdb/cockroach/commit/6dd54b46cc56b7d2b302e0d5ec1509658a1c86f7)? The pull request includes the full code review, which is not relevant for a TA.

Reviewer 4: ```suggestion
[In #123442](https://github.com/cockroachdb/cockroach/commit/6dd54b46cc56b7d2b302e0d5ec1509658a1c86f7), we resolved an issue with CockroachDB in the expiration-to-epoch lease promotion transition process, where the lease's effective expiration could be allowed to regress, resulting in two nodes believing they are the leaseholder for a range.
```

Reviewer 3: This suggestion sounds good to me.

The patch has been applied to maintenance releases of CockroachDB:

- [v23.1.27]({% link releases/v23.1.md%}#v23-1-27)
- [v23.2.11]({% link releases/v23.2.md%}#v23-2-11)
- [v24.1.1]({% link releases/v24.1.md%}#v24-1-1)

This public issue is tracked by [131639](https://github.com/cockroachdb/cockroach/issues/131639).

## Mitigation

Users are encouraged to upgrade to [v23.1.27]({% link releases/v23.1.md%}#v23-1-27), [v23.2.11]({% link releases/v23.2.md%}#v23-2-11), [v24.1.1]({% link releases/v24.1.md%}#v24-1-1), or a later version that includes the patch.

### Detection via logs

> Users can run the script included below to detect part of the symptoms of this technical advisory. The script accepts a [debug.zip]({% link v24.3/cockroach-debug-zip.md %}) and determines whether any lease expiration regressions are evident from the logs.

Reviewer 3: The script to longer accepts a debug.zip. Instead, it must be run from within an unzipped debug.zip directory.

> Users can run the script included below to detect part of the symptoms of this technical advisory. The script accepts a [debug.zip]({% link v24.3/cockroach-debug-zip.md %}) and determines whether any lease expiration regressions are evident from the logs.

Reviewer 4: ```suggestion
Users can run the script [a131639_check.sh](a131639_check.sh) to determine whether any lease expiration regressions are evident from their logs. The script must be run from within a decompressed [debug.zip]({% link v24.3/cockroach-debug-zip.md %}) directory.
```

@nvanbenschoten My suggestion here (1) links to a copy of the script in the docs repo (2) implements your comment (3) removes the language that I found too vague that the script will "detect part of the symptoms of this technical advisory".

I presume users would benefit from a more specific caveat about what the script may or may not find, but LMK what you think, or if this is OK as-is.

As described previously, a lease expiration regression is only one of the two race conditions needed to encounter a lost write, with the other being a stale read during asynchronous intent resolution on the range that had the lease expiration regression. Consequently, this script does not provide a complete method of detection for a lost write. Running this script may result in a limited number of false positives. For complete confirmation, this script should be combined with the method described in [Detection via an audit of primary and secondary indexes](#detection-via-an-audit-of-primary-and-secondary-indexes).

Users should not run the script on a production system. Instead, users should move the `debug.zip` file elsewhere for analysis and run the script there.

The script requires (and asks for) the following dependencies:

- [ripgrep](https://github.com/BurntSushi/ripgrep#installation)
- [DuckDB](https://duckdb.org/docs/installation)

Example usage:

~~~

> ➜ ./ta131639_check.sh debug.prod-crdb2.20240922001746.zip

Reviewer 3: This output is a little stale and also includes more information than we should probably share externally. How about:
```bash
➜ pwd
/path/of/unzipped/debug

➜ /path/to/script/ta131639_check.sh
Searching for liveness epoch increments ...
Searching for slow lease promotions ...
Querying the logs for lease expiration regressions ...

❌ Lease expiration regressions found. Symptoms detected.

symptoms
node 123 observed possible symptoms at 2024-09-01 12:34:56 on ranges 123456
```

  exit 1
fi

# Check if duckdb is installed.
if ! command -v duckdb &> /dev/null; then
  echo "Error: duckdb is not installed. Please install following the instructions in https://duckdb.org/docs/installation"
  exit 1
fi

# Grep the debug.zip for the two queries.
echo "Searching for liveness epoch increments ..."
echo "timestamp,remote_node_id,epoch" > logs_epoch_increment.csv
rg -I -r '$1,$2,$3' 'I([0-9]+ [0-9:.]+).*incremented n(.*) liveness epoch to (.*)' nodes >> logs_epoch_increment.csv || true
if [[ $(wc -l <logs_epoch_increment.csv) -eq 1 ]]; then
    printf "\n✅ No liveness epoch increments found. Symptoms not detected.\n"
    exit 0
fi

echo "Searching for slow lease promotions ..."
echo "timestamp,node_id,range_id,range_span,lease_epoch,lease_proposed,prev_expiration" > logs_slow_lease_promo.csv
rg -I -r '$1,$2,$3,$4,$5,$6,$7' 'W([0-9]+ [0-9:.]+).*,n(\d+),.*,r(\d+)/\d+:(.*),raft\].*client traffic may have been delayed.*epo=(\d+).*pro=(\d+\.\d+).*prev.*exp=(\d+\.\d+).*Request.*' nodes >> logs_slow_lease_promo.csv || true
if [[ $(wc -l <logs_slow_lease_promo.csv) -eq 1 ]]; then
    printf "\n✅ No slow lease promotions found. Symptoms not detected.\n"
    exit 0
fi

> # Query the logs using duckdb.
> echo "Querying the logs for lease expiration regressions ..."
> cat <<EOF > query.sql
> SELECT
>   'node '
>   || node_id
>   || ' observed possible symptoms at '
>   || lease_proposed_min
>   || ' on ranges '
>   || string_agg(range_id, ', ') as symptoms
> FROM
>   (
>     WITH
>       lease_acqs AS (
>         SELECT
>           strptime(timestamp, '%y%m%d %H:%M:%S.%n')::timestamp_ns as timestamp,
>           node_id,
>           range_id,
>           range_span,
>           lease_epoch,
>           timezone('UTC', to_timestamp(lease_proposed)) AS lease_proposed,
>           timezone('UTC', to_timestamp(prev_expiration)) AS prev_expiration
>         FROM 'logs_slow_lease_promo.csv'
>       ),
>       epoch_incs AS (
>         SELECT
>           strptime(timestamp, '%y%m%d %H:%M:%S.%n')::timestamp_ns as timestamp,
>           remote_node_id,
>           epoch
>         FROM 'logs_epoch_increment.csv'
>       )
>     SELECT *, date_trunc('minute', lease_proposed) AS lease_proposed_min
>     FROM lease_acqs JOIN epoch_incs
>     -- where the leaseholder's liveness record had its epoch incremented
>     ON lease_acqs.node_id = epoch_incs.remote_node_id
>     -- and the lease acquisition used an epoch that is less than the one
>     -- that was incremented to
>     AND lease_acqs.lease_epoch < epoch_incs.epoch
>     -- and the lease acquisition began before the epoch increment
>     AND lease_acqs.lease_proposed < epoch_incs.timestamp
>     -- and the lease acquisition completed after the epoch increment
>     AND lease_acqs.timestamp > epoch_incs.timestamp
>     -- and the previous lease's expiration was after the epoch increment
>     AND lease_acqs.prev_expiration > epoch_incs.timestamp
>   )
> GROUP BY node_id, lease_proposed_min
> ORDER BY lease_proposed_min
> EOF
> duckdb -list < query.sql > logs_lease_regressions.txt
> 
> # Output result.
> if [ -s logs_lease_regressions.txt ]; then
>     printf "\n❌ Lease expiration regressions found. Symptoms detected.\n\n"
>     cat logs_lease_regressions.txt
>     exit 1
> else
>     printf "\n✅ No lease expiration regressions found. Symptoms not detected.\n"
>     exit 0
> fi
> ~~~

Reviewer 4: Suggesting that we drop the code, in lieu of the suggested link above

```suggestion
```

### Detection via an audit of primary and secondary indexes

> If you suspect that a particular row may have been subject to a lost write, you can query the primary and secondary indexes to confirm that the row exists in both places.

Reviewer 3: Should we be more prescriptive about why someone might suspect that they were subject to a lost write? Something like:
"If the detection script detects symptoms of a potential lease expiration regression and you suspect that a particular row may have been subject to a lost write ... "

Reviewer 4: @nvanbenschoten This SGTM. Will implement after your re-review. (Sorry, I missed some of your comments as we were both working with the PR at the same time; I think I've now addressed all of them.)
